### DISCUSSION TITLE
How to override stdin for tests simulating user input?

### DISCUSSION THREAD

**Discussion by DavesCodeMusings** [micropython/micropython#9954] [Using MicroPython]
I've got an interactive program that I'd like to write some tests for. It requires simulating user input, so I'm trying to override stdin to pass pre-defined input. Unfortunately, I've succeeded only in crashing the system so far.

I started with this, based on examples I found on the interwebs:
```
import sys
from io import StringIO

with StringIO('Hello World') as f:
    stdin_save = sys.stdin
    sys.stdin = f
    msg = input()
    print('"{}" was passed to input() by overriding stdin.'.format(msg))
    sys.stdin = stdin_save
```

It works great in CPython, but fails in MicroPython. I think I know why, but I'm unsure how to fix it.

In CPython, sys.stdin is a TextIOWrapper
```
>>> import sys
>>> sys.stdin
<_io.TextIOWrapper name='<stdin>' mode='r' encoding='utf-8'>
```

But in MicroPython it's FileIO
```
>>> import sys
>>> sys.stdin
<io.FileIO 0>
```

So I tried changing it:
```
import sys
from io import FileIO

with FileIO('Hello World') as f:
    stdin_save = sys.stdin
    sys.stdin = f
    msg = input()
    print('"{}" was passed to input() by overriding stdin.'.format(msg))
    sys.stdin = stdin_save
```

Now I've made things worse and running the code above causes an unhandled exception that reboots the board. I also tried getting more specific with `sys.stdin.read`, but again I'm just crashing the board.

Not exactly where to go next. I'm hoping someone has done this kind of thing before and can point me in the right direction. Or is there another way to send simulated user input to a MicroPython program's stdin?

**Comment by jimmo** [comment]
@DavesCodeMusings I think you can use `os.dupterm` to achieve this in MicroPython (this is how things like WebREPL inject stdin/stdout). See also https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_uart_repl.py which does the same thing for BLE.

**Reply by DavesCodeMusings** [reply]
Thank you. I will give it a try.
